src/plugins/analysis/file_system_metadata/code/file_system_metadata.py

```diff
@@ -36,85 +58,67 @@ class StatResult(NamedTuple):
     m_time: float
 
 
-class AnalysisPlugin(AnalysisBasePlugin):
+class FileMetadata(BaseModel):
+    mode: str = Field(description="The file's permissions as octal number")
+    mode_human_readable: str = Field(description="The file's permissions as human-readable string (e.g. '-rwxrwxrwx')")
+    name: str = Field(description="The file's name")
+    path: str = Field(description="The file's path")
+    user: str = Field(description="The user name of the file's owner")
+    uid: int = Field(description="The user ID of the file's owner")
+    group: str = Field(description="The group name of the file's owner")
+    gid: int = Field(description="The group ID of the file's owner")
+    modification_time: float = Field(description="The time of the file's last modification (as UNIX timestamp)")
+    access_time: Optional[float] = Field(None, description="The time of the file's last access (as UNIX timestamp)")
+    creation_time: Optional[float] = Field(None, description="The time of the file's creation (as UNIX timestamp)")
+    suid_bit: bool = Field(description='Whether the Setuid bit is set for this file')
+    sgid_bit: bool = Field(description='Whether the Setgid bit is set for this file')
+    sticky_bit: bool = Field(description='Whether the sticky bit is set for this file')
+    key: str = Field(description='Used internally for matching this file in the parent container')
```

Comment:
> Why is mode a string and not an integer?

Octal numbers are IMHO the most intuitive and readable form when looking at file modes (what do you find easier to read? `"755"` or `493`? Please note that we can't store it as `755` because that would be numerically wrong when using the number to check permissions). We could store it as integer and convert it to octal in the frontend but that would not help when looking at the raw data or using the rest endpoints. 

> Does mode_human_readable have to exist? For me that looks like this should be some logic in the frontend.

It doesn't *have* to exist. It could be useful when looking at the raw data but IMHO the `mode` field should be enough for that. I generally don't like to have many filters in the frontend that are only used by one plugin template, though. Maybe we need a way for plugins to specify filters.  

> Do we really need `key`?

`key` is only used internally for matching the results of a container to the files inside (we have this feature that when looking at the results page of the "children", the results from the parent are dynamically loaded so that we have results for individual files because this plugin usually only works on file systems). 

> With this change we have a good change to rethink the schema and I'd like to do that.
Imo we should mirror the [struct stat](https://linux.die.net/man/2/stat) from unix.
It has exactly what the plugin returns plus st_nlink, st_rdev, st_blksize and st_blocks.
If we just mirror this struct exactly then whoever is using the result is most probably familiar with it
and does not have to look into FACT specific things.
I didn't have a look at the fields that are currently missing and if they are implementable, but even if they are not we can leave them out.
What are your thoughts?

It already mostly mirrors stat output because most of the data is generated by stat (or the pathlib equivalent). I intentionally left out the data that I deemed not interesting (or that is not available e.g. for archives/tar) and I don't really see a use for it now (e.g. why would you be interested in the block size?). I also don't really think that using the same key names is really helpful because the readability/usability would be roughly the same for people that are used to work with stat and lower for others (do you really find it easier to work with e.g. "st_mtime" than "modification_time"?). On the other hand, the data contains some additional fields that are not part of stat (e.g. file path, user/group, setuid/setgid bits) which I would like to keep

Reply:
> I generally don't like to have many filters in the frontend that are only used by one plugin template, though. Maybe we need a way for plugins to specify filters

I think that is the wrong way around to thing about frontend & backend.
From my perspective plugins (backend) should just provide raw data and it is the frontends concern to present that data nicely. To exaggerate a bit: Why should we have a frontend if it does nothing but show what we would get  when using the REST api anyways.

> that would not help when looking at the raw data or using the rest endpoints

The latter is not a usecase that I think is useful to consider.
If you use the REST api you most probably want to run some code on what you got from the api, dont you?
So there it would be more useful to provide data in machine-friendly format.

> key is only used internally

Ok I guess.

Reply:
> I think that is the wrong way around to thing about frontend & backend.
From my perspective plugins (backend) should just provide raw data and it is the frontends concern to present that data nicely. To exaggerate a bit: Why should we have a frontend if it does nothing but show what we would get when using the REST api anyways.

That is only partly correct. The plugins have their own templates and are therefore responsible for the presentation of their own results (how could the frontend be responsible for that if you consider that you can add your own plugins with its own templates). If a filter is only used in a template of one plugin, it kinda belongs to that plugin and not the rest of the frontend (it also makes it harder to track what the filter belongs to and ultimately if the filter can be removed at some point if it gets obsolete).